cv_next/app/api/users/fetchUserData/route.ts

```diff
@@ -0,0 +1,36 @@
+"use server";
+
+import SupabaseHelper from "@/server/api/supabaseHelper";
+import logger from "@/server/base/logger";
+import { Tables, ProfileKeys } from "@/lib/supabase-definitions";
+import { NextRequest, NextResponse } from "next/server";
+
+export async function POST(req: NextRequest) {
```

Comment:
Here for example we must have some sort of user validation

Reply:
This apparently is a bit problomatic
If we want to have RLS on the profiles that checks if someone is active, how is it that the active status is in the profiles table?

Reply:
I see what you mean; but we have 0auth authentication right? I forgot, what's our auth flow? This here should be available after a sign-in, so essentially we should have some 0auth creds we can check against (jwt or something similar).
This is what I mean by the user validation.

Reply:
I see what you mean, I can add a simple test just to see that we have a JWT but checking the JWT's content will lead to the error that the API routes solved...

Reply:
Wait, walk me through the flow. We get login through oauth, and what claims do we get from there? What info do we have that we later use in the "user-handling" process? Username? UserId?
I have an idea that is not very smart but can at least help us bypass the stupid supabase bullshit. We can create a simple list file on the deployment server (not in supabase) that dups the relation we need to match the user and their access level. It'd be referable if we can use uuids when checking against this list. It should be updated either in a cron job, or straight via the user-handling actions.
It's a silly data duplication that might have issues, but we have to handle this somehow. We can't leave these routes unguarded..

Reply:
1. Login using oAuth of google
2. Getting a code from google and exchanging it to get an active session from supabase
3. In the cookies we see several things:
- Access Token which is encrypted with a key we have in supabase. The decrypted version has all of the other things that we can find in the cookie.
- User-related stuff like id, email, name, photo, etc.
4. For each supabase request, we send the cookie which includes the id. Using supabase's RLS rules we can get the id, see the "user_type" which specifies the permissions we gave the user that made the request as we don't want anyone that logs in to simply have everything.
5. If the user is at least "active", it can have read access to everything but can't edit things like cvs that aren't his, or comments, etc.
The full list is in another issue if you want to see it.
What I suggest is this:
1. Create a separate table for relations between userid and the perms which will be update with a trigger similar to how we create the user's entry in our current table.
2. Set the RLS of that table to only allow select to everyone.
3. Each RLS of our tables will see if the user is in that table as active or above

Reply:
And the problem with supabase that made us go to routes? In which step is it? 4?
Wait, the RLS rules are applied on all user requests? i.e. if a request is sent to this route (and all routes), perhaps we won't know if the user is allowed to perform this action, but supabase will not allow it?
If that's the case, perhaps we can simply rely on the supabase RLS rules, and only correctly handle the errors received from it.
(and if that's what you meant to do in the first place, sorry I didn't understand it)

Also I've checked our RLS - as I understand it, it draws from "whitelisted" and "admins" views - "whitelisted" currently is all the users and not active and above. Just an FYI, you probably know.


Oh, and preferably catching unauthorized requests as early as possible is what should guide us, but the other option is to basically provide free access to all user accounts (even if it's just ids and access level), which I'm not sure is better.

Reply:
We switched to routes to api using server actions so we could ignore them in the middleware so technically yes 4
RLS works on every access to the DB - in the backend we don't check perms because supabase does it for us with RLS.
whitelisted is active and above we just dont have any inactive users
The problem with catching it as soon as possible is that to see unauthorized requests we need to see who is unauthorized which means getting the data of the profiles i.e. id and type so we need to allow public select access to this
The main problem is that I couldn't get the correct RLS to allow me to get the profile data on profiles that weren't me without full access
If full access:
Got everything
With checking if I'm whitelisted:
Got only myself

Reply:
Wait, why do you need more than yourself? Also, can you confirm the behavior I talked about - that currently all of our routes will get blocked in supabase level via the RLS if the user is not active and above? Which means that (albeit not the best solution, as we've talked about early-as-possible) we can rely on the RLS and simply catch the errors here, which means we don't "need" another level of security on the route-level (as I initially thought).

Actually on second thought, perhaps making the profile uuid-type data public might not be _that_ risky (uuids are already essentially public, and user type doesn't need to be secure). I'll admit I'm not that nifty on the security side so asking Alik might be good. Perhaps I'll ask Niki as well. What do you think?

Reply:
We need others to see their uid to see data about them for the profile/full name of commenter or uploader
Yes as of right now each path will get blocked at supabase level according to user type(active and above). The only exception is as mentioned the profiles which allows select to everyone. Other than the profiles we can trust rls.
Asking them can be a good idea.
Something I just remembered is that profiles also contain if that person is looking for a job and on which category which might be a bit unsafe…

Reply:
On the first point - you mean on the normal flow when viewing e.g. a cv page, user with "active" type (i.e. non-admin) will get rejected for requests on profiles for comments? Why would that happen? The profiles table is defined for whitelisted in the RLS, no?
And I'll try asking them we'll see what they have to say.

Reply:
Currently we have RLS public true for selecting profiles
If I change it I can only see myself(I won't be able to fetch other usernames or ids and stuff like that
Whitelisted is only for the id and type
Notice that there's a policy for public which is just true

Reply:
Apparently you were correct about the whitelist thing, updated it